### PR TITLE
Improvement: Improve workspace state 

### PR DESCRIPTION
**Pull request by bbaa77770**
LU-1708

### CODE REVIEW

File: src/app/ui/widgets/WorkspaceVisualizer/Visualizer.jsx [hunk 1]
```diff
@@ -514,17 +514,17 @@ class Visualizer extends PureComponent {
...
             const size = nextProps.size;
             if (nextProps.preview) {
-                this.previewPrintableArea.updateSize({
+                this.previewPrintableArea && this.previewPrintableArea.updateSize({
                     x: size.x * 2,
                     y: size.y * 2
                 });
             } else {
-                this.state.printableArea.updateSize({
+                this.previewPrintableArea && this.state.printableArea.updateSize({
```
Comment (xiekun1992) [inline src/app/ui/widgets/WorkspaceVisualizer/Visualizer.jsx:522]: this.previewPrintableArea =>this.state.printableArea

File: src/app/flux/workspace/index.js [hunk 2]
```diff
@@ -21,6 +21,11 @@ export const WORKSPACE_STAGE = {
 };
 
 const INITIAL_STATE = {
+    headType: '',
+    toolHead: '',
```
Comment (meiliang1123) [inline src/app/flux/workspace/index.js:25]: workspace uses connected machine, should save headtype in flux/machine

File: src/app/ui/widgets/Connection/SerialConnection.jsx [hunk 3]
```diff
@@ -95,25 +100,47 @@ function SerialConnection() {
...
+        let machineToolHead = '';
+        // TODO
+        if (machineHeadType === HEAD_LASER) {
+            machineToolHead = LEVEL_ONE_POWER_LASER_FOR_SM2;
+            dispatch(workspaceActions.updateMachineState({
+                headType: HEAD_LASER,
+                toolHead: LEVEL_ONE_POWER_LASER_FOR_SM2,
+                series: machineSeries
+            }));
+        } else if (machineHeadType === '10w-laser') {
```
Comment (meiliang1123) [inline src/app/ui/widgets/Connection/SerialConnection.jsx:114]: split machineHeadType and machineToolHead

File: src/app/ui/widgets/Marlin/Marlin.jsx [hunk 4]
```diff
@@ -49,7 +51,7 @@ function MarlinWidget({ widgetActions }) {
 
     return (
         <div>
-            {headType === HEAD_PRINTING && <Printing />}
+            {(headType === HEAD_PRINTING || headType === '3dp') && <Printing />}
```
Comment (meiliang1123) [inline src/app/ui/widgets/Marlin/Marlin.jsx:54]: why headType value has 2 value, should fix this bug.

**Review by meiliang1123 — CHANGES_REQUESTED** [review]

**Pull request merged**